File: internal/flowengine/syncconfig.go
package flowengine

import (
	"encoding/json"
	"strconv"
	"strings"

	"github.com/gnasdev/gn-drive/internal/store"
)

// ProfileFromSyncConfig is the single flow-run flag path: Operation.sync_config
// JSON (snake_case or camelCase keys, same as frontend lib/syncConfig.ts) maps
// onto store.Profile, which syncengine turns into rclone.ProfileFlags.
// Profile columns used only as a flag DTO for flow path — not as workspace unit.
func ProfileFromSyncConfig(raw json.RawMessage) *store.Profile {
	return profileFromSyncConfig(raw)
}

// profileFromSyncConfig maps a flow Operation.sync_config JSON bag onto a
// store.Profile used as the flag carrier for StartPathSync.
func profileFromSyncConfig(raw json.RawMessage) *store.Profile {
	p := &store.Profile{Parallel: 4}
	if len(raw) == 0 {
		return p
	}
	var m map[string]any
	if err := json.Unmarshal(raw, &m); err != nil {
		return p
	}

	if v, ok := asBool(m, "dry_run", "dryRun"); ok {
		p.DryRun = v
	}
	if v, ok := asInt(m, "parallel"); ok && v > 0 {
		p.Parallel = v
	}
	if v, ok := asInt(m, "bandwidth"); ok && v > 0 {
		p.Bandwidth = v
	}
	if v, ok := asInt(m, "multi_thread_streams", "multiThreadStreams"); ok && v > 0 {
		p.MultiThreadStreams = intPtr(v)
	}
	if v, ok := asString(m, "buffer_size", "bufferSize"); ok {
		p.BufferSize = v
	}
	if v, ok := asInt(m, "retries"); ok && v > 0 {
		p.Retries = intPtr(v)
	}
	if v, ok := asInt(m, "low_level_retries", "lowLevelRetries"); ok && v > 0 {
		p.LowLevelRetries = intPtr(v)
	}
	if v, ok := asString(m, "max_duration", "maxDuration"); ok {
		p.MaxDuration = v
	}
	if v, ok := asBool(m, "check_first", "checkFirst"); ok {
		p.CheckFirst = v
	}
	if v, ok := asString(m, "order_by", "orderBy"); ok {
		p.OrderBy = v
	}
	if v, ok := asString(m, "retries_sleep", "retriesSleep"); ok {
		p.RetriesSleep = v
	}
	if v, ok := asFloat(m, "tps_limit", "tpsLimit"); ok && v > 0 {
		p.TpsLimit = floatPtr(v)
	}
	if v, ok := asString(m, "conn_timeout", "connTimeout"); ok {
		p.ConnTimeout = v
	}
	if v, ok := asString(m, "io_timeout", "ioTimeout"); ok {
		p.IoTimeout = v
	}

	if v, ok := asStringSlice(m, "included_paths", "includedPaths"); ok {
		p.IncludedPaths = v
	}
	if v, ok := asStringSlice(m, "excluded_paths", "excludedPaths"); ok {
		p.ExcludedPaths = v
	}
	if v, ok := asString(m, "min_size", "minSize"); ok {
		p.MinSize = v
	}
	if v, ok := asString(m, "max_size", "maxSize"); ok {
		p.MaxSize = v
	}
	if v, ok := asString(m, "max_age", "maxAge"); ok {
		p.MaxAge = v
	}
	if v, ok := asString(m, "min_age", "minAge"); ok {
		p.MinAge = v
	}
	if v, ok := asInt(m, "max_depth", "maxDepth"); ok && v > 0 {
		p.MaxDepth = intPtr(v)
	}
	if v, ok := asString(m, "filter_from_file", "filterFromFile"); ok {
		p.FilterFromFile = v
	}
	if v, ok := asString(m, "exclude_if_present", "excludeIfPresent"); ok {
		p.ExcludeIfPresent = v
	}
	if v, ok := asBool(m, "use_regex", "useRegex"); ok {
		p.UseRegex = v
	}
	if v, ok := asBool(m, "delete_excluded", "deleteExcluded"); ok {
		p.DeleteExcluded = v
	}

	if v, ok := asInt(m, "max_delete", "maxDelete"); ok && v > 0 {
		p.MaxDelete = intPtr(v)
	}
	if v, ok := asBool(m, "immutable"); ok {
		p.Immutable = v
	}
	if v, ok := asString(m, "max_transfer", "maxTransfer"); ok {
		p.MaxTransfer = v
	}
	if v, ok := asString(m, "max_delete_size", "maxDeleteSize"); ok {
		p.MaxDeleteSize = v
	}
	if v, ok := asString(m, "suffix"); ok {
		p.Suffix = v
	}
	if v, ok := asBool(m, "suffix_keep_extension", "suffixKeepExtension"); ok {
		p.SuffixKeepExtension = v
	}
	if v, ok := asString(m, "backup_path", "backupPath"); ok {
		p.BackupPath = v
	}

	if v, ok := asBool(m, "size_only", "sizeOnly"); ok {
		p.SizeOnly = v
	}
	if v, ok := asBool(m, "update_mode", "updateMode"); ok {
		p.UpdateMode = v
	}
	if v, ok := asBool(m, "ignore_existing", "ignoreExisting"); ok {
		p.IgnoreExisting = v
	}

	if v, ok := asString(m, "delete_timing", "deleteTiming"); ok {
		p.DeleteTiming = v
	}

	if v, ok := asString(m, "conflict_resolution", "conflictResolution"); ok {
		p.ConflictResolution = v
	}
	if v, ok := asBool(m, "resilient"); ok {
		p.Resilient = v
	}
	if v, ok := asString(m, "max_lock", "maxLock"); ok {
		p.MaxLock = v
	}
	if v, ok := asBool(m, "check_access", "checkAccess"); ok {
		p.CheckAccess = v
	}
	if v, ok := asString(m, "conflict_loser", "conflictLoser"); ok {
		p.ConflictLoser = v
	}
	if v, ok := asString(m, "conflict_suffix", "conflictSuffix"); ok {
		p.ConflictSuffix = v
	}

	return p
}

func intPtr(v int) *int          { return &v }
func floatPtr(v float64) *float64 { return &v }

func asBool(m map[string]any, keys ...string) (bool, bool) {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			switch t := v.(type) {
			case bool:
				return t, true
			}
		}
	}
	return false, false
}

func asInt(m map[string]any, keys ...string) (int, bool) {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			switch t := v.(type) {
			case float64:
				return int(t), true
			case int:
				return t, true
			case string:
				if n, err := strconv.Atoi(strings.TrimSpace(t)); err == nil {
					return n, true
				}
			}
		}
	}
	return 0, false
}

func asFloat(m map[string]any, keys ...string) (float64, bool) {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			switch t := v.(type) {
			case float64:
				return t, true
			case int:
				return float64(t), true
			case string:
				if n, err := strconv.ParseFloat(strings.TrimSpace(t), 64); err == nil {
					return n, true
				}
			}
		}
	}
	return 0, false
}

func asString(m map[string]any, keys ...string) (string, bool) {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			if s, ok := v.(string); ok && strings.TrimSpace(s) != "" {
				return strings.TrimSpace(s), true
			}
		}
	}
	return "", false
}

func asStringSlice(m map[string]any, keys ...string) ([]string, bool) {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			switch t := v.(type) {
			case []any:
				out := make([]string, 0, len(t))
				for _, x := range t {
					if s, ok := x.(string); ok && strings.TrimSpace(s) != "" {
						out = append(out, strings.TrimSpace(s))
					}
				}
				return out, true
			case []string:
				return append([]string(nil), t...), true
			}
		}
	}
	return nil, false
}

File: frontend/src/lib/syncConfig.ts
/**
 * Operation.sync_config JSON shape for flow runs.
 * Backend maps this via flowengine.ProfileFromSyncConfig (snake_case or camelCase keys)
 * then syncengine → rclone flags. Keep keys aligned with that single mapping path.
 */
import { normalizeFlowAction, type FlowAction } from '@/constants/forms'

export interface SyncConfig {
  action: FlowAction

  // Performance
  parallel?: number
  bandwidth?: number // MB/s
  multiThreadStreams?: number
  bufferSize?: string
  retries?: number
  lowLevelRetries?: number
  maxDuration?: string
  checkFirst?: boolean
  orderBy?: string
  retriesSleep?: string
  tpsLimit?: number
  connTimeout?: string
  ioTimeout?: string

  // Filtering
  includedPaths?: string[]
  excludedPaths?: string[]
  minSize?: string
  maxSize?: string
  maxAge?: string
  minAge?: string
  maxDepth?: number
  filterFromFile?: string
  excludeIfPresent?: string
  useRegex?: boolean
  deleteExcluded?: boolean

  // Safety
  dryRun?: boolean
  maxDelete?: number
  immutable?: boolean
  maxTransfer?: string
  maxDeleteSize?: string
  suffix?: string
  suffixKeepExtension?: boolean
  backupPath?: string

  // Comparison
  sizeOnly?: boolean
  updateMode?: boolean
  ignoreExisting?: boolean

  // Sync-specific (push)
  deleteTiming?: 'before' | 'during' | 'after' | ''

  // Bisync
  conflictResolution?: string
  resilient?: boolean
  maxLock?: string
  checkAccess?: boolean
  conflictLoser?: string
  conflictSuffix?: string
}

function pickNum(raw: Record<string, unknown>, ...keys: string[]): number | undefined {
  for (const k of keys) {
    const v = raw[k]
    if (typeof v === 'number' && !Number.isNaN(v)) return v
    if (typeof v === 'string' && v.trim() !== '' && !Number.isNaN(Number(v))) return Number(v)
  }
  return undefined
}

function pickStr(raw: Record<string, unknown>, ...keys: string[]): string | undefined {
  for (const k of keys) {
    const v = raw[k]
    if (typeof v === 'string' && v.trim() !== '') return v
  }
  return undefined
}

function pickBool(raw: Record<string, unknown>, ...keys: string[]): boolean | undefined {
  for (const k of keys) {
    const v = raw[k]
    if (typeof v === 'boolean') return v
  }
  return undefined
}

function pickStrArr(raw: Record<string, unknown>, ...keys: string[]): string[] | undefined {
  for (const k of keys) {
    const v = raw[k]
    if (Array.isArray(v)) {
      return v.filter((x): x is string => typeof x === 'string')
    }
  }
  return undefined
}

/** Normalize raw sync_config JSON into a typed SyncConfig. */
export function parseSyncConfig(raw: unknown, fallbackAction: string = 'push'): SyncConfig {
  const o =
    raw && typeof raw === 'object' && !Array.isArray(raw)
      ? (raw as Record<string, unknown>)
      : {}
  const action = normalizeFlowAction(
    (typeof o.action === 'string' && o.action) || fallbackAction,
  )
  const cfg: SyncConfig = { action }

  const bag = cfg as unknown as Record<string, unknown>
  const n = (camel: keyof SyncConfig, ...keys: string[]) => {
    const v = pickNum(o, ...keys)
    if (v !== undefined) bag[camel as string] = v
  }
  const s = (camel: keyof SyncConfig, ...keys: string[]) => {
    const v = pickStr(o, ...keys)
    if (v !== undefined) bag[camel as string] = v
  }
  const b = (camel: keyof SyncConfig, ...keys: string[]) => {
    const v = pickBool(o, ...keys)
    if (v !== undefined) bag[camel as string] = v
  }
  const a = (camel: keyof SyncConfig, ...keys: string[]) => {
    const v = pickStrArr(o, ...keys)
    if (v !== undefined) bag[camel as string] = v
  }

  n('parallel', 'parallel')
  n('bandwidth', 'bandwidth')
  n('multiThreadStreams', 'multiThreadStreams', 'multi_thread_streams')
  s('bufferSize', 'bufferSize', 'buffer_size')
  n('retries', 'retries')
  n('lowLevelRetries', 'lowLevelRetries', 'low_level_retries')
  s('maxDuration', 'maxDuration', 'max_duration')
  b('checkFirst', 'checkFirst', 'check_first')
  s('orderBy', 'orderBy', 'order_by')
  s('retriesSleep', 'retriesSleep', 'retries_sleep')
  n('tpsLimit', 'tpsLimit', 'tps_limit')
  s('connTimeout', 'connTimeout', 'conn_timeout')
  s('ioTimeout', 'ioTimeout', 'io_timeout')

  a('includedPaths', 'includedPaths', 'included_paths')
  a('excludedPaths', 'excludedPaths', 'excluded_paths')
  s('minSize', 'minSize', 'min_size')
  s('maxSize', 'maxSize', 'max_size')
  s('maxAge', 'maxAge', 'max_age')
  s('minAge', 'minAge', 'min_age')
  n('maxDepth', 'maxDepth', 'max_depth')
  s('filterFromFile', 'filterFromFile', 'filter_from_file')
  s('excludeIfPresent', 'excludeIfPresent', 'exclude_if_present')
  b('useRegex', 'useRegex', 'use_regex')
  b('deleteExcluded', 'deleteExcluded', 'delete_excluded')

  b('dryRun', 'dryRun', 'dry_run')
  n('maxDelete', 'maxDelete', 'max_delete')
  b('immutable', 'immutable')
  s('maxTransfer', 'maxTransfer', 'max_transfer')
  s('maxDeleteSize', 'maxDeleteSize', 'max_delete_size')
  s('suffix', 'suffix')
  b('suffixKeepExtension', 'suffixKeepExtension', 'suffix_keep_extension')
  s('backupPath', 'backupPath', 'backup_path')

  b('sizeOnly', 'sizeOnly', 'size_only')
  b('updateMode', 'updateMode', 'update_mode')
  b('ignoreExisting', 'ignoreExisting', 'ignore_existing')

  s('deleteTiming', 'deleteTiming', 'delete_timing')

  s('conflictResolution', 'conflictResolution', 'conflict_resolution')
  b('resilient', 'resilient')
  s('maxLock', 'maxLock', 'max_lock')
  b('checkAccess', 'checkAccess', 'check_access')
  s('conflictLoser', 'conflictLoser', 'conflict_loser')
  s('conflictSuffix', 'conflictSuffix', 'conflict_suffix')

  return cfg
}

/** Persist shape: prefer snake_case for Go store/profile alignment + camel for Wails. */
export function serializeSyncConfig(cfg: SyncConfig): Record<string, unknown> {
  const out: Record<string, unknown> = { action: cfg.action }
  const set = (snake: string, camel: string, v: unknown) => {
    if (v === undefined || v === null || v === '') return
    if (Array.isArray(v) && v.length === 0) return
    out[snake] = v
    out[camel] = v
  }
  set('parallel', 'parallel', cfg.parallel)
  set('bandwidth', 'bandwidth', cfg.bandwidth)
  set('multi_thread_streams', 'multiThreadStreams', cfg.multiThreadStreams)
  set('buffer_size', 'bufferSize', cfg.bufferSize)
  set('retries', 'retries', cfg.retries)
  set('low_level_retries', 'lowLevelRetries', cfg.lowLevelRetries)
  set('max_duration', 'maxDuration', cfg.maxDuration)
  set('check_first', 'checkFirst', cfg.checkFirst)
  set('order_by', 'orderBy', cfg.orderBy)
  set('retries_sleep', 'retriesSleep', cfg.retriesSleep)
  set('tps_limit', 'tpsLimit', cfg.tpsLimit)
  set('conn_timeout', 'connTimeout', cfg.connTimeout)
  set('io_timeout', 'ioTimeout', cfg.ioTimeout)
  set('included_paths', 'includedPaths', cfg.includedPaths)
  set('excluded_paths', 'excludedPaths', cfg.excludedPaths)
  set('min_size', 'minSize', cfg.minSize)
  set('max_size', 'maxSize', cfg.maxSize)
  set('max_age', 'maxAge', cfg.maxAge)
  set('min_age', 'minAge', cfg.minAge)
  set('max_depth', 'maxDepth', cfg.maxDepth)
  set('filter_from_file', 'filterFromFile', cfg.filterFromFile)
  set('exclude_if_present', 'excludeIfPresent', cfg.excludeIfPresent)
  set('use_regex', 'useRegex', cfg.useRegex)
  set('delete_excluded', 'deleteExcluded', cfg.deleteExcluded)
  set('dry_run', 'dryRun', cfg.dryRun)
  set('max_delete', 'maxDelete', cfg.maxDelete)
  set('immutable', 'immutable', cfg.immutable)
  set('max_transfer', 'maxTransfer', cfg.maxTransfer)
  set('max_delete_size', 'maxDeleteSize', cfg.maxDeleteSize)
  set('suffix', 'suffix', cfg.suffix)
  set('suffix_keep_extension', 'suffixKeepExtension', cfg.suffixKeepExtension)
  set('backup_path', 'backupPath', cfg.backupPath)
  set('size_only', 'sizeOnly', cfg.sizeOnly)
  set('update_mode', 'updateMode', cfg.updateMode)
  set('ignore_existing', 'ignoreExisting', cfg.ignoreExisting)
  set('delete_timing', 'deleteTiming', cfg.deleteTiming)
  set('conflict_resolution', 'conflictResolution', cfg.conflictResolution)
  set('resilient', 'resilient', cfg.resilient)
  set('max_lock', 'maxLock', cfg.maxLock)
  set('check_access', 'checkAccess', cfg.checkAccess)
  set('conflict_loser', 'conflictLoser', cfg.conflictLoser)
  set('conflict_suffix', 'conflictSuffix', cfg.conflictSuffix)
  return out
}

export function defaultSyncConfig(action: FlowAction = 'push'): SyncConfig {
  return { action }
}

/** Short chips for view mode when non-default options are set. */
export function syncConfigSummaryChips(cfg: SyncConfig): string[] {
  const chips: string[] = []
  if (cfg.dryRun) chips.push('dry-run')
  if (cfg.parallel && cfg.parallel > 0) chips.push(`×${cfg.parallel}`)
  if (cfg.bandwidth && cfg.bandwidth > 0) chips.push(`${cfg.bandwidth}M`)
  if (cfg.includedPaths?.length) chips.push(`+${cfg.includedPaths.length} include`)
  if (cfg.excludedPaths?.length) chips.push(`−${cfg.excludedPaths.length} exclude`)
  if (cfg.maxAge) chips.push(`max-age ${cfg.maxAge}`)
  if (cfg.minSize || cfg.maxSize) chips.push('size filter')
  if (cfg.conflictResolution) chips.push(`conflict:${cfg.conflictResolution}`)
  if (cfg.immutable) chips.push('immutable')
  if (cfg.sizeOnly) chips.push('size-only')
  return chips
}
